keps/sig-api-machinery/2019-02-13-custom-resource-definitions.md

```diff
@@ -0,0 +1,212 @@
+
+When it became clear that ThirdPartyResources were not only viable, but actually the predominant way
+to extend the Kubernetes API Server. Core API Machinery maintainers renamed ThirdPartyResources to
+CustomResourceDefinitions and rewrote most of the core code to improve it's compatability and
+integration with a significantly more mature API server.
+
+At this point, the core CustomResource mechanisms have matured to a point where there is discussion
+about how to build out core required APIs as custom resources (see below).
+
+## Proposal
+
+### User Stories
+
+#### Story 1: Easy extensibility
+
+Alice is a developer who runs her own Kubernetes cluster. Alice mostly deploys Java applications. One
+day Alice thinks: "Instead of always using Kubernetes `Deployment` objects, I'd like to create a `JavaDeployment` object that abstracts lots of the boilerplate of the `PodTemplate` inside the `Deployment`
+object into a `JavaPodTemplate` that knows how to run Java applications.
+
+Alice uses CustomResourceDefinitions and a simple controller and extends her cluster to have 
+`JavaDeployments` in a single afternoon.
+
+#### Story 2: Lightweight cluster add-ons
+
+Jane is a cluster adminstrator. Here users are hopeless at deploying and managing certificates as
+Secrets in Kubernetes. Jane wants to add a `RotatingSecret` which uses the API Server's signing 
+capabilities to automatically rotate certificates for her users. Jane fines an open source project
+that has this capability, and Jane installs the extension into the clusters she administers without
+worrying about extra storage components that she needs to monitor and manage.
+
+#### Story 3: An ecosystem of plugin providers.
+
+Julia is running a startup. She wants to make it easy for the world to perform XSS testing against
+their websites. Julia builds a custom piece of software that does XSS attacks, and also builds a
+`XSSDetector` custom resource. Julia can build her business based on people installing her software
+as an extension inside of their Kubernetes clusters no matter where they are running.
+
+### Implementation details
+
+For details of how to use `CustomResourceDefinitions` there is excellent documentation [elsewhere](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/).
+
+### Patterns for Custom Resources
+
+There are three main patterns for custom resources, ordered in complexity from simplest to most complex:
+   * Dumb data objects
+   * Compiler patterns
+   * Operator patterns
+
+#### Dumb data objects
+
+Dumb data objects are the simplest use case for CustomResourceDefinitions. In this case, the resource
+is simply used for CRUD operations. A user might create an object with `kubectl` and an application
+might read that object in when it starts and periodically resynchronize that data during it's operation.
+The object is used only for data storage, there is no control-plane machinery running in the cluster.
```

Comment:
In the non-goals it's noted that:

> Develop an arbitrary key-value store for applications. Custom resources are control-plane objects. 

Does the dumb data object details act more like a key-value store for an app? This notes there  is no control-plane part. Is this case a non-goal?